### PR TITLE
[Fix][Master] Fix master-server execution logic thread unsafe problem

### PR DESCRIPTION
**Pull request by zhuangchong**
## Verify this pull request

This pull request is code cleanup without any test coverage.

*(or)*

This pull request is already covered by existing tests, such as *(please describe tests)*.

(or)

This change added tests and can be verified as follows:

### CODE REVIEW

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/consumer/TaskPriorityQueueConsumer.java [hunk 1]
```diff
@@ -120,7 +121,7 @@ public void run() {
      * batch dispatch with thread pool
      */
     private List<TaskPriority> batchDispatch(int fetchTaskNum) throws TaskPriorityQueueException, InterruptedException {
-        List<TaskPriority> failedDispatchTasks = new ArrayList<>();
+        List<TaskPriority> failedDispatchTasks = Collections.synchronizedList(new ArrayList<>());
```
Comment (caishunfeng) [inline dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/consumer/TaskPriorityQueueConsumer.java:124]: This is the funtion execute in the single thread, so is it necessary to use the `Collections.synchronizedList`? 
Will it affect performance?

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/MasterSchedulerService.java [hunk 2]
```diff
@@ -179,7 +180,8 @@ private void scheduleProcess() throws Exception {
     }
 
     private List<ProcessInstance> command2ProcessInstance(List<Command> commands) {
-        List<ProcessInstance> processInstances = new ArrayList<>(commands.size());
+
+        List<ProcessInstance> processInstances = Collections.synchronizedList(new ArrayList<>(commands.size()));
```
Comment (caishunfeng) [inline dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/MasterSchedulerService.java:184]: The same question.

**Review by caishunfeng — COMMENTED** [review]
BTW, Is there a bug issue related to this PR?

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/consumer/TaskPriorityQueueConsumer.java (hunk 1 above)
Comment (zhuangchong) [reply dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/consumer/TaskPriorityQueueConsumer.java:124]: In this method, the thread pool consumerThreadPoolExecutor operates on the variable failedDispatchTasks collection.

image

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessService.java [hunk 3]
```diff
@@ -886,11 +886,12 @@ private ProcessInstance constructProcessInstance(Command command, String host) {
...
         if (processInstance.getCommandParam() != null) {
             Map<String, String> processCmdParam = JSONUtils.toMap(processInstance.getCommandParam());
-            for (Map.Entry<String, String> entry : processCmdParam.entrySet()) {
-                if (!cmdParam.containsKey(entry.getKey())) {
-                    cmdParam.put(entry.getKey(), entry.getValue());
+            processCmdParam.forEach((k, v) -> {
+                String value = cmdParam.get(k);
+                if (StringUtils.isBlank(value)) {
+                    cmdParam.put(k, v);
                 }
```
Comment (ruanwenjun) [inline dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessService.java]:
```suggestion
```

**Review by ruanwenjun — COMMENTED** [review]
Good job, just a little might need to change, I think use `containsKey` or `computeIfAbsent` is better than check the value is null.

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/consumer/TaskPriorityQueueConsumer.java (hunk 1 above)
Comment (ruanwenjun) [reply dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/consumer/TaskPriorityQueueConsumer.java:124]: Yes, this list is used in multiple threads, it's needed to use threadsafe collection here.

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessService.java (hunk 3 above)
Comment (zhuangchong) [reply dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessService.java]: done

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/MasterSchedulerService.java (hunk 2 above)
Comment (zhuangchong) [reply dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/MasterSchedulerService.java:184]: In this method, the thread pool masterPrepareExecService operates on the variable processInstances collection.

image

**Review by ruanwenjun — APPROVED** [review]
LGTM

**Review by caishunfeng — APPROVED** [review]
+1

**Pull request merged**